xarray/core/coordinates.py

```diff
@@ -964,22 +964,32 @@ def __getitem__(self, key: Hashable) -> T_DataArray:
         return self._data._getitem_coord(key)
 
     def _update_coords(
-        self, coords: dict[Hashable, Variable], indexes: Mapping[Any, Index]
+        self, coords: dict[Hashable, Variable], indexes: dict[Hashable, Index]
     ) -> None:
         coords_plus_data = coords.copy()
         coords_plus_data[_THIS_ARRAY] = self._data.variable
         dims = calculate_dimensions(coords_plus_data)
-        if not set(dims) <= set(self.dims):
-            raise ValueError(
-                "cannot add coordinates with new dimensions to a DataArray"
-            )
-        self._data._coords = coords
 
-        # TODO(shoyer): once ._indexes is always populated by a dict, modify
-        # it to update inplace instead.
-        original_indexes = dict(self._data.xindexes)
-        original_indexes.update(indexes)
-        self._data._indexes = original_indexes
+        if set(dims) > set(self.dims):
+            for k, v in coords.items():
+                if any(d not in self.dims for d in v.dims):
+                    # allow any coordinate associated with an index that shares at least
+                    # one of dataarray's dimensions
+                    temp_indexes = Indexes(
+                        indexes, {k: v for k, v in coords.items() if k in indexes}
+                    )
+                    if k in indexes:
+                        index_dims = temp_indexes.get_all_dims(k)
+                        if any(d in self.dims for d in index_dims):
+                            continue
```

Comment:
I don't understand this at all. Can we make it simpler? Or at least expand the comments. The repeated creation of `Indexes` seems like it can have a large impact.